### DISCUSSION TITLE
How to return binary stream and use input

### DISCUSSION THREAD

**Discussion by lanwin** [RobinTail/express-zod-api#1305] [Q&A] [answered]
Hi, I am currently trying out the wonderful little package and trying to figure out how to return a binary stream and use and input object. 

My example is an pdf to image converter where the request body is the pdf and formatting parameters are passed via query. The result is an binary image.

From what I read in the docs I should use an EndpointsFactory with positive and negative response for that. But then how to use an zod input object with that?

**Comment by RobinTail** [comment, marked as answer]
Hello, @lanwin .

In order to receive the whole request body as a binary stream, see raw data processing:
https://github.com/RobinTail/express-zod-api#accepting-raw-data

If possible, consider also regular file uploads instead:
https://github.com/RobinTail/express-zod-api#file-uploads

In order to stream the image in response, see the example of custom ResultHandler:
https://github.com/RobinTail/express-zod-api#non-object-response

> But then how to use an zod input object with that?

- For endpoint input schema use `ez.raw()` which is an alias for `z.object({ raw: z.file().buffer() })` — you will need to adjust you config with `rawParser` option in order to make this little trick work. This feature is available in the recent v14.2.0. See details in Readme and Changelog.
- For endpoint output schema, use `z.object({ ... })` with a simple property, like filename, that can then be handled by a custom ResultHandler for reading and streaming that file. If it's not possible, use `z.object({ stream: z.file().buffer() })` right away, and make the custom ResultHandler to forward that `stream` property as a response.

I hope I managed to help you. Please let me know if something need to be clarified.
